packages/open-next/src/adapters/http/responseStreaming.ts

```diff
@@ -0,0 +1,181 @@
+import http from 'node:http';
+import type { ResponseStream } from '../types/aws-lambda.js';
+import { debug, error } from '../logger.js';
+import { getString, headerEnd } from './utils.js';
+
+const HEADERS = Symbol();
+const BODY = Symbol();
+
+export class StreamingServerResponse extends http.ServerResponse {
+  [HEADERS]: Record<string, string> = {};
+  [BODY]: any[] = [];
+  responseStream: ResponseStream;
+  fixHeaders: (headers: Record<string, string>) => void;
+  private _wroteHeader = false;
+  private _done = false;
+  private _hasWritten = false;
+  timer: NodeJS.Timer | null = null;
+
+  get headers() {
+    return this[HEADERS];
+  }
+
+  setHeader(name: string, value: string | number | readonly string[]): this {
+    // @ts-ignore
+    this[HEADERS][name.toLowerCase()] = value;
+    return this;
+  }
+
+  writeHead(
+    statusCode: number,
+    statusMessage?: string | undefined,
+    headers?: http.OutgoingHttpHeaders | http.OutgoingHttpHeader[] | undefined
+  ): this;
+  writeHead(
+    statusCode: number,
+    headers?: http.OutgoingHttpHeaders | http.OutgoingHttpHeader[] | undefined
+  ): this;
+  writeHead(
+    statusCode: unknown,
+    statusMessage?: unknown,
+    headers?: unknown
+  ): this {
+    if (!this._wroteHeader) {
```

Comment:
can we go the other way around? aka. not to nest when not necessary?

I would suggest going `if(this._wroteHeader) { return this }`